src/mem/freelist.h

```diff
@@ -0,0 +1,297 @@
+#pragma once
+/**
+ * This file encapsulates the in disused object free lists
+ * that are used per slab.
+ */
+
+#include "../ds/address.h"
+#include "../ds/cdllist.h"
+#include "../ds/dllist.h"
+#include "../ds/helpers.h"
+#include "allocconfig.h"
+
+#include <iostream>
+
+namespace snmalloc
+{
+#ifdef CHECK_CLIENT
+  /**
+   * The key that is used to encode free list pointers.
+   * This should be randomised at startup in the future.
+   */
+  inline static uintptr_t global_key =
+    static_cast<size_t>(bits::is64() ? 0x9999'9999'9999'9999 : 0x9999'9999);
+
+  /**
+   * Used to turn a location into a key.  This is currently
+   * just the value of the previous location + 1.
+   */
+  inline static uintptr_t initial_key(void* p)
+  {
+    return address_cast(p) + 1;
+  }
+#endif
+
+  static inline bool different_slab(uintptr_t p1, uintptr_t p2)
+  {
+    return ((p1 ^ p2) >= SLAB_SIZE);
+  }
+
+  static inline bool different_slab(uintptr_t p1, void* p2)
+  {
+    return different_slab(p1, address_cast(p2));
+  }
+
+  static inline bool different_slab(void* p1, void* p2)
+  {
+    return different_slab(address_cast(p1), address_cast(p2));
+  }
+
+  /**
+   * Free objects within each slab point directly to the next.
+   * The next_object pointer can be encoded to detect
+   * corruption caused by writes in a UAF or a double free.
+   */
+  class FreeObject
+  {
+    FreeObject* next_object;
+
+    static FreeObject* encode(uintptr_t local_key, FreeObject* next_object)
+    {
+#ifdef CHECK_CLIENT
+      if constexpr (aal_supports<IntegerPointers>)
+      {
+        // Simple symmetric encoding
+        auto next = address_cast(next_object);
+        constexpr uintptr_t MASK = bits::one_at_bit(bits::BITS / 2) - 1;
+        // Mix in local_key
+        auto key = local_key ^ global_key;
+        next ^= (((next & MASK) + 1) * key) & ~MASK;
+        next_object = reinterpret_cast<FreeObject*>(next);
+      }
+#else
+      UNUSED(local_key);
+#endif
+      return next_object;
+    }
+
+  public:
+    static FreeObject* make(void* p)
```

Comment:
Come to think of it, should we be applying cache-friendly offsets _here_ (and removing below) so that the `Allocator` functions can just pass around `void *`s that always point to the start of the object?

Reply:
The only issue is that they use allocator local state to introduce the offsets.

Reply:
Leaving as can be part of future PR. If we decide to keep cache-friendly offsets.